### PR TITLE
Increases customizability of activation functions

### PR DESCRIPTION
**Pull request by craabreu**
## Description
Accepts a `torch.nn.Module` object as the activation function when constructing a message passing or predictor object.

## Examples

Library usage:
```py
from chemprop import nn
import torch
message_passing = nn.BondMessagePassing(activation=torch.nn.Softplus())
ffn = nn.RegressionFFN(activation=torch.nn.Softplus())
```

CLI usage:
```bash
chemprop train \
    --activation Softplus --activation-args beta=1 threshold=20 \
    --data-path tests/data/regression.csv \
    --task-type regression \
    --output-dir train_example
```

## Questions

Is the implemented CLI syntax adequate?

## Relevant issues
Addresses #878.

## Checklist
- [x] linted with flake8
- [x] unit tests added
- [x] CLI options implemented
- [x] CLI tests for interlayer activation
- [x] ~added output activation option~
- [x] ~added unit tests for output activation~
- [x] ~CLI tests for output activation~
- [x] dropped support for SELU activation function

### CODE REVIEW

**Comment by craabreu** [comment]
Possible solution for the CLI:

1. Include a new choice `--activation CLASS`, where `CLASS` is any class defined in `torch.nn.modules.activation`.

2. Include a new argument `--activation-argument VALUE/NAME=VALUE` with `action='append'` as a way to pass positional and/or non-default keyword arguments to the class constructor.

**Comment by craabreu** [comment]
This is the currently implemented CLI syntax:

```
$ chemprop train -h
  ...
  --activation {RELU,LEAKYRELU,PRELU,TANH,SELU,ELU,CUSTOM}
                        Activation function in message passing/FFN layers
  ...
  --activation-module {Threshold,ReLU,RReLU,Hardtanh,ReLU6,Sigmoid,Hardsigmoid,Tanh,SiLU,Mish,Hardswish,ELU,CELU,SELU,GLU,GELU,Hardshrink,LeakyReLU,LogSigmoid,Softplus,Softshrink,MultiheadAttention,PReLU,Softsign,Tanhshrink,Softmin,Softmax,Softmax2d,LogSoftmax}
                        A CUSTOM activation function for message passing/FFN layers.
  --activation-args ACTIVATION_ARGS
                        Comma-separated arguments for a CUSTOM activation function.
   ...
```

**Comment by JacksonBurns** [comment]
This is not related to the changes in this PR but this is a good place to mention this:

I just realized that we expose the Self-normalizing Exponential Linear Unit (SELU) activation function, which requires using AlphaDropout in place of regular Dropout and also requires a [different initialization of linear weights](https://pytorch.org/docs/stable/generated/torch.nn.SELU.html#selu) (which we don't do). We should probably remove SELU, or else implement these requirements.

**Comment by JacksonBurns** [comment]
@craabreu we had a feature request in the linked Discussion post - could you add support for a customizable `--output-activation`, i.e. the ability to pass something like `Sigmoid` to be used on the output of the FNN?

**Comment by craabreu** [comment]
@JacksonBurns, do you mean adding an extra activation block, possibly a different one, here?

https://github.com/chemprop/chemprop/blob/1b8f4133b092fb0241f690f93a2b11e7db07c6dd/chemprop/nn/ffn.py#L34-L55

**Comment by JacksonBurns** [comment]
Yes, I think that would work!

**Comment by craabreu** [comment]
I simplified how the activation function is specified in the CLI and edited https://github.com/chemprop/chemprop/pull/1146#issuecomment-2596350793 accordingly for clarity.

**Comment by craabreu** [comment]
Added option to append an FFN predictor with an extra activation function, as described in https://github.com/chemprop/chemprop/pull/1146#issuecomment-2634581895.

The CLI specification works as follows:

```
FFN args:
  ...
  --output-activation {NONE,RELU,CELU,ELU,GELU,GLU,HARDSHRINK,HARDSIGMOID,HARDSWISH,HARDTANH,LEAKYRELU,LOGSIGMOID,LOGSOFTMAX,MISH,MULTIHEADATTENTION,PRELU,RRELU,RELU6,SELU,SILU,SIGMOID,SOFTMAX,SOFTMAX2D,SOFTMIN,SOFTPLUS,SOFTSHRINK,SOFTSIGN,TANH,TANHSHRINK,THRESHOLD}
                        Activation function in the output layer of the FFN top model.
  --output-activation-args [OUTPUT_ACTIVATION_ARGS ...]
                        Arguments for the output activation function (Example: arg1 arg2 key1=value1 key2=value2).

```

**Comment by craabreu** [comment]
This PR is ready for review.

**Comment by GZ82** [comment]
@JacksonBurns thanks for your suggestion, I have forward my comments from discussion to this PR. Here is my comments:
this is probably possible if [change source code](https://chemprop.readthedocs.io/en/latest/tutorial/python/activation.html)
but I think it will be more user friendly if can add an arg e.g., `activation_arg = 'Sigmoid' for classification`
in class like ` chemprop.nn.predictors.RegressionFFN`

**Review by KnathanM — CHANGES_REQUESTED** [review]
I don't see the need for `output_activation` in the predictors. The example use case of adding a sigmoid to `RegressionFFN` to have it work for classification is not needed because we already have `BinaryClassificationFFN`. But also importantly in our classification FFN, the sigmoid is not applied during training because the BCELoss function expects logits, not probabilities. My vote is to remove `output_activation`. 

I also note that these changes don't allow using a different activation for the message passing and predictor via the CLI, but that is fine. If someone really has a use case for that, they can use Chemprop from their own python script.

**Comment by KnathanM** [comment]
> This is not related to the changes in this PR but this is a good place to mention this:
> 
> I just realized that we expose the Self-normalizing Exponential Linear Unit (SELU) activation function, which requires using AlphaDropout in place of regular Dropout and also requires a [different initialization of linear weights](https://pytorch.org/docs/stable/generated/torch.nn.SELU.html#selu) (which we don't do). We should probably remove SELU, or else implement these requirements.

In the Self-Normalizing Neural Network paper (linked in the torch.nn.SELU page) , they call SELU "scaled exponential linear unit". In this PR, all activation functions from torch are made available. I don't think we need to specifically exclude SELU from the available option just because it is normalized used in conjunction with other things to make a Self-Normalizing Neural Network. 

But we could still exclude it from the default search space of our hyperparameter optimization. I don't how big of a change it is to adjust the default search space of activation functions, so maybe we wait to change that until 2.2?

**Comment by JacksonBurns** [comment]
The `output_activation` was added per user request. Is there a way that we could only do the `output_activation` during inference (I think this is probably the main ask).

**Comment by craabreu** [comment]
I also vote to remove the output activation option. Most users will never need this functionality, and those who need it can do it via API by passing a custom `output_transform` when instantiating a `RegressionFFN`. A possibility is to expand the API from this:

https://github.com/chemprop/chemprop/blob/44dbdf9a742eabed0213eacc0290952281b1bc7c/chemprop/nn/predictors.py#L119

To this:
```py
output_transform: UnscaleTransform | nn.Module | None = None,
```

Does that make sense?

**Review by JacksonBurns — CHANGES_REQUESTED** [review]
Looks good to me. Few organizational edits, and:
 - we should add a demo of how to set `Sigmoid` as the `output_transform` for an FFN, since we aren't adding `output_activation` (which to be clear, I agree with that decision in hindsight)
 - `SELU` should be removed from the supported activation functions. When using `SELU` we are supposed to initialize linear layers differently (see https://pytorch.org/docs/stable/generated/torch.nn.SELU.html#selu). Because we don't, it seems misleading to make it an option. As a user, I would expect that the software would do this correctly behind-the-scenes, but it doesn't.

**Comment by craabreu** [comment]
@JacksonBurns Unwrapping `Linear` from `Sequential` in the code line below does not fundamentally change the model, but ruins backward compatibility with checkpoint files. For this reason, I will keep the line unchanged.

https://github.com/chemprop/chemprop/blob/1b8f4133b092fb0241f690f93a2b11e7db07c6dd/chemprop/nn/ffn.py#L46

**Comment by JacksonBurns** [comment]
@craabreu good catch! Is this ready for another review?

**Comment by JacksonBurns** [comment]
> Yes, @JacksonBurns. It's ready for another review. I didn't include a tutorial on applying an activation function to the FFN's output. This became off-topic and can have its own PR, in my opinion. Does that make sense?

Yes, that's fine

File: chemprop/nn/utils.py [hunk 1]
```diff
@@ -10,25 +10,30 @@ class Activation(EnumMapping):
...
     Returns
     -------
     nn.Module
         The activation function module.
     """
+    if isinstance(activation, nn.Module):
+        if isinstance(activation, nn.modules.activation.SELU):
+            raise RuntimeError(
+                "SELU activation (intended for self-normalizing networks) is not supported."
+            )
```
Comment (JacksonBurns) [inline chemprop/nn/utils.py:35]:
```suggestion
            raise TypeError(
                "Support for SELU activation (intended for self-normalizing networks) has been removed in v2.2.0"
            )
```

Comment (JacksonBurns) [reply chemprop/nn/utils.py:35]: I think this is slightly clearer

**Review by JacksonBurns — APPROVED** [review]
Small suggestion (ignore if you choose) about making an error different but otherwise this LGTM

**Comment by JacksonBurns** [comment]
@KnathanM this needs a re-review from you

File: chemprop/cli/utils/args.py [hunk 2]
```diff
@@ -32,3 +32,19 @@ def uppercase(x: str):
...
+    if s in {"true", "false"}:
+        v = s == "true"
+    try:
+        v = int(s)
+    except ValueError:
+        try:
+            v = float(s)
+        except ValueError:
+            v = s
+    return {k[0]: v} if k else v
```
Comment (KnathanM) [inline chemprop/cli/utils/args.py:50]: This needs an "else" block. Otherwise if `s=true`, then `v=True`, but then it will get to `v=s` and override `v` with `"true"`.

File: docs/source/tutorial/python/activation.ipynb
Comment (KnathanM) [inline docs/source/tutorial/python/activation.ipynb:1]: I think we should clarify what we mean by "The following activation functions are available in Chemprop." Maybe "The following activation functions can be specified using their names, i.e. `activation="relu"`."

File: docs/source/tutorial/python/activation.ipynb [hunk 3]
```diff
@@ -68,54 +68,84 @@
...
+      "Tanhshrink\n",
+      "Softmin\n",
+      "Softmax\n",
+      "Softmax2d\n",
+      "LogSoftmax\n"
+     ]
+    }
+   ],
+   "source": [
+    "for activation in torch.nn.modules.activation.__all__:\n",
```
Comment (KnathanM) [inline docs/source/tutorial/python/activation.ipynb:117]: I don't know how much value there is in printing out all the activation modules. I think it is sufficient to just say you can pass any torch Module.

**Review by KnathanM — COMMENTED** [review]
I want to summarize the changes around the `SELU` activation function. This PR:

* Removes SELU from the default hpopt search space for activation functions.
* "SELU" is excluded from `_ACTIVATION_FUNCTIONS`, which people from using `--activation selu` on the command line to train a model with SELU.
* The `get_actvation_function` used in the mp and ffn constructors will throw a `TypeError` if a user tries to build a model with `SELU`. 

I believe this would break any models that use SELU, forcing the users to retrain with a different activation function or stay on an old chemprop version. Because of this backwards incompatibility, I think we should rethink what our goal is. 

It sounds like our main goal is to prevent user confusion that could arise from having SELU listed along side other activation function like ReLU and Tanh. The users may incorrectly think that using SELU is enough to get a self normalizing NN. Towards this goal, I agree with removing SELU from our hpopt search space and am neutral about removing it from the choices for `--activation`. 

I don't think that we need to disallow SELU from the model level though. If someone specifically passes in an `nn.SELU` instance as the activation function for their mp or ffn, it isn't our job to stop them. In the end, SELU is just a function like any other function. While its main use case is in self normalizing NNs, that doesn't have to be its only use case. 

Additionally older v2 models could have `activation='selu'` in their checkpoint files. For these models to load correctly we would need to continue to allow `activation='selu'` in `get_activation_function`. If "selu" is not in `Activation` (because we remove it from the hpopt search space), we need to add an `if activation == "selu"; return nn.SELU()` block, possibly with a comment about how that is there to maintain backwards compatibility.

**Comment by JacksonBurns** [comment]
@KnathanM agreed re: backwards compat. We can leave SELU in the codebase, but at minimum we should remove it from the search space for hpopt. I'm also in favor of adding a warning when `--activation selu` is passed that using SELU doesn't actually make chemprop self-normalizing.

File: docs/source/tutorial/cli/train.rst [hunk 4]
```diff
@@ -130,7 +130,7 @@ Model performance is often highly dependent on the hyperparameters used. Below i
  * :code:`--message-hidden-dim <n>` Hidden dimension of the messages in the MPNN (default 300)
  * :code:`--depth <n>` Number of message-passing steps (default 3)
  * :code:`--dropout <n>` Dropout probability in the MPNN & FFN layers (default 0)
- * :code:`--activation <activation_type>` The activation function used in the MPNN and FNN layers. Options include :code:`relu`, :code:`leakyrelu`, :code:`prelu`, :code:`tanh`, :code:`selu`, and :code:`elu`. (default :code:`relu`)
+ * :code:`--activation <activation_type>` The activation function used in the MPNN and FNN layers. Options include :code:`relu`, :code:`leakyrelu`, :code:`prelu`, :code:`tanh`, and :code:`elu`. (default :code:`relu`)
```
[the remark this answers is not in the document]
Comment (craabreu) [reply docs/source/tutorial/cli/train.rst:133]:
```suggestion
 * :code:`--activation <activation_type>` The activation function used in the MPNN and FNN layers. Run :code:`chemprop train -h` to see a list of activation functions supported via CLI.
```

**Review by JacksonBurns — APPROVED** [review]

File: chemprop/cli/utils/args.py [hunk 5]
```diff
@@ -32,3 +32,20 @@ def uppercase(x: str):
 
 def lowercase(x: str):
     return x.lower()
+
+
+def activation_function_argument(argument):
+    """Parse a command line option for an activation function argument."""
+    *k, s = argument.split("=", 1)
+    s = s.strip().lower()
```
[the remark this answers is not in the document]
Comment (craabreu) [reply chemprop/cli/utils/args.py:40]: None of the CLI options accepts string-valued arguments, but it's worth preparing the function for future additions.

File: chemprop/cli/utils/args.py [hunk 6]
```diff
@@ -32,3 +32,20 @@ def uppercase(x: str):
...
+    if s in {"true", "false"}:
+        v = s == "true"
```
Comment (craabreu) [inline chemprop/cli/utils/args.py:42]:
```suggestion
    s = s.strip()
    if s.lower() in {"true", "false"}:
        v = s.lower() == "true"
```

File: chemprop/cli/utils/parsing.py [hunk 7]
```diff
@@ -455,3 +456,15 @@ def parse_indices(idxs):
                 indices.append(int(idx))
         return indices
     return idxs
+
+
+def parse_activation(cls: type[nn.Module], arguments: list | None) -> nn.Module:
```
[the remark this answers is not in the document]
Comment (craabreu) [reply chemprop/cli/utils/parsing.py:461]: Good catch. I'll follow your suggestion.

**Review by KnathanM — APPROVED** [review]
